pyth[![Contributor Covenant](https://img.shields.io/badge/Contributor%20Covenant-2.1-4baaaa.svg)](code_of_conduct.md)


# McHoRD
Python package for determining homologous recombination status from capture panels


## About

McHoRD is a packaged designed to use genome wide SNP/CNV capture panels to determine the homologous recombination status
of tumour samples without a germline sample for reference.

It works on a similar principle to genome wide loss of heterozygosity scores, but as the status of the SNPs is unknown it
uses the distributional structure of heterozygous SNPs as compared to a previously defined reference population.
In testing on whole genome sequencing samples it provided consistent scores to HRDetect and CHORD.

As the distribution and its departure in HRD cases will differ depending on the capture assay sites selected and the 
population assessed cut of scores will need to be determined by the user to ensure they are appropriate for their use
case. As a guide in our WGS based testing scores above 0.25 were classified as HRD, with only HPV driven tumours
registering high scores without a known HR defect. (This is biologically plausible and under investigation).

## Installation

Survival volume is written for use with Python3
Version 0.2 requires Python 3.9 or higher.

It has dependency on:
- scipy >= 1.6

To install from the github repository

```pip3 install git+https://github.com/PapenfussLab/McHoRD.git```

To run the tests use

```python3 -m mchord.tests.test_mchord```

## Usage

McHoRD is provided as is for research use only under the BSD 3 Clause license.
Any express or implied warranties of merchantability or fitness for a particular purpose are disclaimed.

```
McHoRD - determining HRD status from genome wide SNP capture panels

positional arguments:
  infiles

optional arguments:
  -h, --help            show this help message and exit
  --version             print version information and exit
  --make_null           generate a null distribution
  --null_distribution NULL_DISTRIBUTION
  --targets TARGETS
  --show_ascii
  --outfile [OUTFILE]

```

McHoRD requires a bed file of capture locations (which must be approximately evenly spaced across the genome) and a VCF
file of SNP calls. At this time only samples called with Octopus (Cooke, 2021) in somatic mode are supported.
Extremes of tumour purity and FFPE can severely impact accuracy and appropriate QC cut offs should be determined
specifically for your end to end workflow.

To create a reference distribution you will require a collection of known HR competent samples processed in the same
manner as for your test samples.

A reference distribution is created and stored in a JSON format file with the following command

```
mchord --make_null --outfile FILE_TO_SAVE_TO.json --targets YOUR_CAPTURE.bed OCTOPUS_CALLS.vcf 
```

Once you have a null distribution a score is generated by executing the following command

```
mchord --targets YOUR_CAPTURE.bed --null_distribution FILE_FROM_MAKENULL_COMMAND.json OCTOPUS_CALLS.vcf
```

## Contributing to McHoRD
McHoRD is licensed under the BSD-3-Clause license.  You are free to fork this repository under the terms of that license.
If you have suggested changes please start by raising an issue in the issue tracker.
Pull requests are welcome and will be included at the discretion of the author.
Bug reports should be made to the issue tracker.
Difficulty in understanding how to use the software is a documentation bug, and should also be raised on the
issue tracker so your question and my response are easily found by others.

McHoRD aims to maintain a respectful and inclusive community and adopts the
[contributor covenant v2.1](code_of_conduct.md)

## Citing McHoRD

McHoRD is currently unpublished. Please contact the author if you intend to publish using this tool and an appropriate
version will be deposited with a permanent DOI suitable for citation.

## References
@article{octopus,
   author = {Cooke, Daniel P. and Wedge, David C. and Lunter, Gerton},
   title = {A unified haplotype-based method for accurate and comprehensive variant calling},
   journal = {Nature Biotechnology},
   ISSN = {1546-1696},
   DOI = {10.1038/s41587-021-00861-3},
   url = {https://doi.org/10.1038/s41587-021-00861-3
https://www.nature.com/articles/s41587-021-00861-3.pdf},
   year = {2021},
   type = {Journal Article}
}